### DISCUSSION TITLE
Annotation reuse feedback

### DISCUSSION THREAD

**Discussion by pdawyndt** [dodona-edu/feedback-and-support#35] [General]
The first time I tried to save and reuse a comment, I already hit the limitation that comments currently cannot be reused across exercises (and I guess also not across courses or course admins). But sometimes comments have to potential to be reused broader than within the same exercise. Something like general tips & tricks or code smells, for example, as are given by linters. So we might extend our scope of reusing annotations and maybe use the different contexts (exercises, course, course admin) to rank the search results.

My [use case](https://dodona.ugent.be/en/submissions/11455797/) was exactly such a generic comment that I wanted to reuse across exercises:

![image](https://user-images.githubusercontent.com/5736113/193201265-41866dcb-1b2c-4696-af61-947c060cf318.png)

**Comment by pdawyndt** [comment]
I'm also not convinced that the approach to save a comment in anticipation of it being reused in the future is the best possible approach. I believe it would be more logical that I could search across all comments (so then we really need a good search and ranking mechanism) and only at the time when a comment (or a variantion thereof) is reused for the first time we could model it as a shared comment that has an extra title and links to all instances where it has been used.

**Reply by pdawyndt** [reply]
I confirm that upon first use of the "reuse annations" feature, I'm missing the possibility to search for annotations that I haven't saved explicitly, without the need to go back to previous submissions in orde to save that earlier annotation for reuse.

**Comment by chvp** [comment]
It's not possible to use the `Shift+Enter` shortcut from the saved annotation title input, which slows me down a little.

**Reply by pdawyndt** [reply]
Would also be nice if you could search for a saved annotation and use the arrow up/down keys to move through the suggested options.

**Reply by jorg-vr** [reply]
`shift+enter` fixed by https://github.com/dodona-edu/dodona/pull/4358

**Comment by chvp** [comment]
When reusing/searching for annotations, it would be useful if they are sorted by longest match in the title (and then by longest match in the content?).

**Reply by jorg-vr** [reply]
Currently the matching is exact. Thus all matches will be the same length.

Can I rephrase your question as: Sort matches in the title before those that only match in content?

**Comment by toonijn** [comment]
When re-editing an edited reused comment, the edits are forgotten. (as discussed with @jorg-vr @chvp)

**Reply by jorg-vr** [reply]
Probably caused by the searchfield looking up the reused and pasting that in the input field. (This is expected behaviour when the searchfield is eddited, but should not happen on first open of the edit view of a comment)

**Reply by jorg-vr** [reply]
fixed by https://github.com/dodona-edu/dodona/pull/4023 or https://github.com/dodona-edu/dodona/pull/4083

**Comment by pdawyndt** [comment]
I would find it more useful if the list of "saved comments" in the right margin were ordered such that the comment that is added/used most recently is displayed first. Happens more often that I want to modify a comment immediately after I have stored it. Alternative would be to support multiple options for ordering we could choose from.

**Comment by pdawyndt** [comment]
If an annotation has been reused (or stored for reuse), reviewers could also see the title of the reused annotation together with the text of the annotation itself (I'm not convinced yet that the title would have added value for students) instead of only showing a "link"-icon. Making the title a hyperlink could be used to update/access the "master annotation".

Here are some options for displaying it

![image](https://user-images.githubusercontent.com/5736113/201358419-1b5b06e8-e1ff-43f4-9aa8-c64abc53ba03.png)

![image](https://user-images.githubusercontent.com/5736113/201358650-ceb7a67e-b7cb-4b46-91c0-e3814898b5fc.png)

**Comment by pdawyndt** [comment]
I feel there are times when you only want to update the "master annotation", only the "reused version" or both. We need to think how these different options are provided.

- make visible if the text of an annotation deviates from its master version
- show both the text of an annotation and the text of its annotation side-by-side (diffed) and allow to overwrite one with the other (in both directions; or more advaced merges as in git diff viewers)

**Comment by pdawyndt** [comment]
Allow versions of master annotations in both English and Dutch: both title and text. Having the option to autotranslate from one language to another as a starting point for having translations in both languages would be a nice to have.

**Reply by pdawyndt** [reply]
This could be extended to have annotations inlined into a source file that also adjust to the language setting from the user, if taken from a master version that is available in both languages.

**Comment by pdawyndt** [comment]
Support Markdown in the titles of reusable annotations.
![image](https://user-images.githubusercontent.com/5736113/201364847-9698dc16-53bd-42db-9182-7c28959273e7.png)

**Comment by pdawyndt** [comment]
If we provide the option to save an annotation for later reuse, that we also have the option to remove it from the list of saved annotations.

**Reply by jorg-vr** [reply]
When editing a saved annotation, you have a button to delete it
Does this satisfy your requirement?

**Comment by pdawyndt** [comment]
When adding a new annotation, the field for reusing a save document comes on top, but the field to enter the text for a new annotation below it is the one who gets the first focus. This is not logical, as the top field should get the focus. Maybe the option to reuse a saved annotation should move to the bottom.

![image](https://user-images.githubusercontent.com/5736113/201471482-c4ccd492-332d-4828-b46a-b00423d05e7e.png)

Also, when pressing the `Tab` key in an empty "Reuse a saved comment" field, a reused annotation is selected although no previous annotations are displayed when the field is empty. This is unexpected behaviour.

**Comment by pdawyndt** [comment]
Sharing reusable annotations among multiple reviewers would definitely have been an asset now that we have reviewed the same assignments with multiple reviewers.

**Reply by pdawyndt** [reply]
In addition, I would also have been able to reuse some generic annotations among the two assignments I have been reviewing. Actually, I already have compiled a list of generic annotations in a Google Doc, that I have been reusing across multiple review sessions.

**Comment by pdawyndt** [comment]
Table of saved annotations](https://dodona.ugent.be/saved_annotations) is too wide. As a consequence, the action button to edit a saved annotation is hidden at the right margin.

![image](https://user-images.githubusercontent.com/5736113/201473927-0d8eebea-dfdb-4f9a-b766-c167d5f11fab.png)

**Reply by pdawyndt** [reply]
Would be nice if this table could be sorted by clicking the column headers.

**Reply by pdawyndt** [reply]
Also support Markdown for rendering the `Title` and the `Text` in this table.

**Reply by pdawyndt** [reply]
Add search filters for `User`, `Course` and `Exercise`

**Reply by pdawyndt** [reply]
Hyperlink the number of reuses to a list of all submissions where the annotation has been reused. This could be used to go back to those instances (and make changes if needed; e.g. update the annotation based on an update of the master version).

**Comment by pdawyndt** [comment]
Condition that saved annotations must have a unique title seems to restrictive. Especially if we start to share annotations across reviewers.

**Reply by pdawyndt** [reply]
When trying so save an annotation that uses a title that was already assigned to another saved annotation, an error message is displayed. After dismissing the error message (dialog), the annotation is there, but is not saved for reused (unexpected behaviour).

**Comment by pdawyndt** [comment]
What are your global conclusions after using the new features for reusing comments while reviewing student submissions?

**Reply by pdawyndt** [reply]
FYI: I used the new features to assess 1/3 of the submissions for two different assignments.

**Global conclusion**

In the past, I used Google Docs to keep track of comments added during review for later reuse. I consider the options for reusing comments that are integrated in Dodona are 100% replacement for my use of Google Docs. Actually, for this review session I had anticipated to save my comments both in Google Docs and Dodona for later reuse. I stopped doing the double work after having assessed twenty-something submissions, as it turned out I could always fetch saved comments from Dodona and I didn't need to comments in my Google Doc.

The listing of all saved comments in my Google Doc still gives a better overview (grouped per function), but my short-time memory turned out to be good enough to recall that I could reuse a save comment.

Although I did use the list of save features from time to time, in those cases I always used it from the separate page (where the features can be filtered) rather than from the listing in the right margin of the page where the assessment is made.

**Most often missed features**

- translations of comments in both Dutch and English (in most cases I now used Dutch and hope students will ask for a translation, where in the past I was making more translations if I thought this would be needed based on the name of the student)
- use saved comments across assignments (I still have a list of generic comments in my Google Doc that I have compiled during previous review sessions, and I keep reusing these comments; now I have to copy them from my Google Doc for each assignment)

**Comment by niknetniko** [comment]
In general, the feature works fine. My biggest "complaint" is that the current workflow requires me to know beforehand if I will want to re-use an annotation or not. Especially at the start of grading a new exercise, this is often not clear to me: sometimes I think an issue will come up frequently and it doesn't, or the reverse: multiple students unexpectedly (to me at least) make the same mistake.

I have thought a bit on how this might be solved with what we currently have. Perhaps we should save the last N different annotations (e.g. N = 20-50) automatically. In the annotation text field, we could then provide auto-completion similar to an IDE for those saved annotations (e.g. with some fuzzy search library). So when you start typing an annotation that is sufficiently similar to one of the N last ones, we show a suggestion to use that text. To reduce server load, this might be done completely in the client (e.g. the server just sends the last N annotations). I think this would complement the saved annotations nicely: after using the same annotations a few times, e.g. 2-3, we could automatically suggest to the user they might want to save the annotation in question.

**Comment by toonijn** [comment]
When the cursor is halfway inside a comment, page-up scrolls the page down.

**Comment by niknetniko** [comment]
I am always inclined to add the title of my annotation I want to save into the search field (probably since the search field is located where the title normally goes in other applications). However, making the search field do double duty as search field and title field will probably be confusing and hard to get right.

Thus, I was thinking that it would be better if the title field was always shown. Additionally, this would allow us to tick the checkbox for saving the annotation automatically if one types something in the title field. This is one click less.

**Comment by rien** [comment]
I personally miss the ability to update previous annotations in one go if desired ("Overwrite all linked annotations"). The main use case would be for spelling mistakes, or if the annotator thinks of a better way to explain a certain mistake.

This should only change the annotations that exactly match the original annotation though. This way one can an annotation "unlink" from the original annotation. this could even be made clear in the UI by adding an explicit "unlock" / "unlink" button.

**Reply by rien** [reply]
This was not added to avoid confusion about what would happen with annotations that are different from the original annotation.
